lisa/target.py

```diff
@@ -211,61 +210,112 @@ class TestEnv(Loggable, HideExekallID):
     freeze when using :meth:`freeze_userspace`.
     """
 
-    def __init__(self, target_conf:TargetConf, plat_info:PlatformInfo=None, res_dir:ArtifactPath=None):
+    CONF_CLASS = TargetConf
+    INIT_KWARGS_KEY_MAP = {
+        'devlib_excluded_modules': ['devlib', 'excluded-modules']
+    }
+
+    def __init__(self, kind, name='<noname>', tools=[], res_dir=None,
+        plat_info=None, workdir=None, device=None, host=None, port=None,
+        username=USERNAME_DEFAULT, password=None, keyfile=None, devlib_platform=None,
+        devlib_excluded_modules=[]
+    ):
+
         super().__init__()
         logger = self.get_logger()
 
-        board_name = target_conf.get('name')
-        if not res_dir:
-            name = board_name or type(self).__qualname__
-            time_str = datetime.now().strftime('%Y%m%d_%H%M%S.%f')
-            name = '{}-{}'.format(name, time_str)
-            res_dir = os.path.join(LISA_HOME, RESULT_DIR, name)
+        self.name = name
 
-        # That res_dir is for the exclusive use of TestEnv itself, it must not
-        # be used by users of TestEnv
-        self._res_dir = res_dir
-        if self._res_dir:
-            os.makedirs(self._res_dir, exist_ok=True)
-            if os.listdir(self._res_dir):
-                raise ValueError('res_dir must be empty: {}'.format(self._res_dir))
+        # That res_dir is for the exclusive use of Target itself, it must not
```

Comment:
Why is that?

Reply:
We don't mix up artifacts anymore. Tests get their own artifact folder when needed, and so does the Target. It should be mentioned that in a typical exekall session, only one Target is created. If tests start to store things there, that boils down to having part of their artifacts stored in the folder dedicated to the first test being run, which is the one for which the Target will be initially created.

Reply:
The reasons the Target has an artifact folder in the first place are:
* GEM5 needs a folder to store its things, since when using GEM5, the Target is not only the connection to the target but the actual virtual board itself.
* RTapp calibration value artifacts, when the Target.plat_info['rtapp']['calib'] is accessed. Since that PlatformInfo instance is deeply linked in a 1-1 fashion with that Target object, it makes sense for it to use a subfolder in the Target's `res_dir`

Reply:
> We don't mix up artifacts anymore. Tests get their own artifact folder when needed, and so does the Target

But then that means we no longer the "all of your experiment results are under that `TestEnv`'s main dir feature" which I found really useful.

Up until recently my results directory was unparsable, but having things stored under that tells me "This is stuff you ran on your HiKey960 and the TestEnv/Target instance was created on X day at Y hour" was really helpful.

Reply:
> But then that means we no longer the "all of your experiment results are under that `TestEnv`'s main dir feature" which I found really useful.

That is already not the case anymore. The ArtifactPathManager is responsible for managing such folder, so the typical case would be to create one per notebook and then call `get_res_dir()` on it to allocate subfolders. One commit in this stack introduces a new-style `TestEnv`, which has its own internal ArtifactPathManager and therefore distributes subfolders this way.

Reply:
So the bottom line is: that feature has definitely nothing to do with connecting to a target, and the new `TestEnv` class is there to centralize some boilerplate that is related to one notebook session.

Reply:
The idea of that series is to remove notebook-specific code that ends up creeping into the lowest level APIs we have. Notebook usecase should be handled as a layer on top of basic entities, by providing a couple of extra classes that are relevant. `exekall` does not have such code sneaking down into `wlgen` for example, it just has a bunch of classes in `exekall_customization` and then use the building blocks of LISA verbatim.